### PR TITLE
PR add-advanced-dialog

### PR DESCRIPTION
**Pull request by DanicaSTFC**
Closes issue #100 

- Add advanced dialog class and tests/example.

### CODE REVIEW

**Comment by DanicaSTFC** [comment]
from terminal, eqt environment active: if running pytests, all pass. 

if set CI=1, then running pythest following errors 
advanced dialog error

File: eqt/ui/FormDialog.py [hunk 1]
```diff
@@ -225,9 +238,9 @@ def saveAllWidgetStates(self):
         '''
         self.formWidget.saveAllWidgetStates()
 
-    def getWidgetStates(self):
+    def getSavedWidgetStates(self):
```
Comment (paskino) [inline eqt/ui/FormDialog.py:241]: add to the CHANGELOG, and add an alias with deprecation as `getWidgetStates`

File: eqt/ui/FormDialog.py [hunk 2]
```diff
@@ -309,3 +322,115 @@ def applyWidgetStates(self, states):
...
+
+        Creates a form dialog and adds a default button to the vertical layout.
+        The default button is located between the form layout and the buttons 'ok' and 'cancel'.
+        The default button is connected to the 'setDefaultValues' method.
+
+        Parameters
+        ----------
+        parent : QWidget or None, optional
+            The parent widget of the advanced form dialog.
+            If None, the dialog is created without a parent.
```
Comment (paskino) [inline eqt/ui/FormDialog.py:340]: parent needs to be a `UIFormWidget`

File: eqt/ui/FormDialog.py [hunk 3]
```diff
@@ -309,3 +322,115 @@ def applyWidgetStates(self, states):
...
+
+        It saves the widget states, calls `onOk`, sets the default widgets
+        to visible and closes the dialog. Adds or updates or removes widgets from the
+        parent based on the values of the widgets in the advanced dialog.
+        """
+        self.saveAllWidgetStates()
+        self.onOk()
+        self.formWidget.setDefaultWidgetStatesVisibleTrue()
+        self.close()
+
```
Comment (paskino) [inline eqt/ui/FormDialog.py:377]:
```suggestion
        super()._onOK()
```

File: eqt/ui/FormDialog.py [hunk 4]
```diff
@@ -309,3 +322,115 @@ def applyWidgetStates(self, states):
...
+                self.dialog_parent.removeWidget(name)
+
+    def setDefaultValues(self):
+        """
+        Sets the widgets in the advanced dialog to their default states.
+
+        Makes the default widget states visible and applies the widget states
+        to the widgets in the form.
+        """
+        self.formWidget.setDefaultWidgetStatesVisibleTrue()
```
Comment (paskino) [inline eqt/ui/FormDialog.py:419]: Can you leave a comment on the reason why this line is necessary?

File: eqt/ui/FormDialog.py [hunk 5]
```diff
@@ -309,3 +322,115 @@ def applyWidgetStates(self, states):
...
+
+        Parameters
+        ----------
+        name : str
+            The name of the widget in the advanced dialog to be displayed in the parent.
+        """
+        if self.dialog_parent is None:
+            raise KeyError('''The advanced-dialog parent is None.
+                           Set the parent if you want to add widgets to it.''')
+        self.display_on_parent_dict[name] = None
```
Comment (paskino) [inline eqt/ui/FormDialog.py:436]: This could be a `list` rather than a `dictionary`.

File: eqt/ui/UIFormWidget.py [hunk 6]
```diff
@@ -458,7 +458,7 @@ def saveAllWidgetStates(self):
         '''
         self.widget_states = self.getAllWidgetStates()
 
-    def getWidgetStates(self):
+    def getSavedWidgetStates(self):
```
Comment (paskino) [inline eqt/ui/UIFormWidget.py:461]: as above, add alias and deprecation warning

File: eqt/ui/FormDialog.py [hunk 7]
```diff
@@ -309,3 +322,115 @@ def applyWidgetStates(self, states):
...
+        """
+        Sets the widgets in the advanced dialog to their default states.
+
+        Makes the default widget states visible and applies the widget states
+        to the widgets in the form.
+        """
+        self.formWidget.setDefaultWidgetStatesVisibleTrue()
+        self.applyWidgetStates(self.formWidget.default_widget_states)
+
+    def addToDictionaryDisplayOnParent(self, name):
```
Comment (paskino) [inline eqt/ui/FormDialog.py:422]: I would rename this method to explain why users would use this method. Also I would remove the term `Dictionary`

```suggestion
    def displayWidgetValueOnParent(self, name):
```

File: examples/advanced_dialog_example.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+        # Example on how to add elements to the FormDialog
+        # add input 1 as QLineEdit
+        qlabel = QtWidgets.QLabel(dialog.groupBox)
+        qlabel.setText("Input 1: ")
+        qwidget = QtWidgets.QLineEdit(dialog.groupBox)
+        qwidget.setClearButtonEnabled(True)
+        # finally add to the form widget
+        dialog.insertWidget(0, 'input_title', QtWidgets.QLabel("Input Values: "))
+        dialog.insertWidget(1, 'input1', qwidget, qlabel)
+        dialog.addToDictionaryDisplayOnParent('input1')
```
Comment (paskino) [inline examples/advanced_dialog_example.py:32]: see suggested rename

File: eqt/ui/FormDialog.py (hunk 1 above)
Comment (DanicaSTFC) [reply eqt/ui/FormDialog.py:241]: done

File: eqt/ui/FormDialog.py (hunk 3 above)
Comment (DanicaSTFC) [reply eqt/ui/FormDialog.py:377]: changed

File: eqt/ui/FormDialog.py (hunk 4 above)
Comment (DanicaSTFC) [reply eqt/ui/FormDialog.py:419]: yes

File: eqt/ui/FormDialog.py (hunk 7 above)
Comment (DanicaSTFC) [reply eqt/ui/FormDialog.py:422]: done

File: eqt/ui/UIFormWidget.py (hunk 6 above)
Comment (DanicaSTFC) [reply eqt/ui/UIFormWidget.py:461]: done

File: test/test__formUI_status_test.py [hunk 9]
```diff
@@ -710,3 +768,143 @@ def test_restoreAllSavedWidgetStates(self):
...
+        self.form_parent.addSpanningWidget(QtWidgets.QPushButton("Open Advanced Dialog"),
+                                           'butt_adv')
+        self.form = AdvancedFormDialog(parent=self.form_parent, title='Advanced form dialog',
+                                       parent_button_name='butt_adv')
+
+        self.layout = self.form.formWidget.uiElements['groupBoxFormLayout']
+        self.vertical_layout = self.form.formWidget.uiElements['verticalLayout']
+        self.add_every_widget()
+        for key in self.list_all_widgets:
+            self.form.addToDictionaryDisplayOnParent(key)
```
[the remark this answers is not in the document]
Comment (DanicaSTFC) [reply test/test__formUI_status_test.py:789]: done

File: eqt/ui/UIFormWidget.py [hunk 10]
```diff
@@ -616,9 +616,9 @@ def saveAllWidgetStates(self):
         '''
         self.widget().saveAllWidgetStates()
 
-    def getWidgetStates(self):
+    def getSavedWidgetStates(self):
```
[the remark this answers is not in the document]
Comment (DanicaSTFC) [reply eqt/ui/UIFormWidget.py:619]: done

File: eqt/ui/FormDialog.py (hunk 5 above)
Comment (DanicaSTFC) [reply eqt/ui/FormDialog.py:439]: done

**Review by paskino — APPROVED** [review]

**Pull request merged**